Pin Repositories to a Specific Git Commit

The RPM should use a specific commit instead of building from the latest master. This will help properly version the RPM and to provide a stable/usable package.

Latest commits as of today:
* 2018-02-04 PublicEngine 77739dfa35d9ab5d13f965c62f275080282d8a29
* 2018-09-30 Core3 e655ca90ec270139ad7200e0cd15ec511950e90e

When running through a build on Fedora 28 with GCC 8.3.1, a handful of class-memaccess warnings appear which result in the build failing.

```
(MMOCoreORB)$ make -j 4 build-cmake
...
/home/user/rpmbuild/BUILD/Core3-c8872ec05aae83066271cc7226c0eb6899b936e5/MMOEngine/include/system/util/ArrayList.h:633:33: error: 'void* realloc(void*, size_t)' moving an object of non-trivially copyable type 'class sys::thread::LockableTrace'; use 'new' and 'delete' instead [-Werror=class-memaccess]
       elementData = (E*) realloc(elementData, (elementCapacity = newCapacity) * sizeof(E));
                          ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
cc1plus: all warnings being treated as errors
```

This should be addressed by adding this argument to the build:

```
-Wno-error=class-memaccess
```

--- a/README.md
+++ b/README.md
@@ -6,12 +6,12 @@
 
 # SWGEmu Server Packages
 
-This repository aims to provide easily available and unofficial Linux packages for creating and installing a Star Wars Galaxies Emulator (SWGEmu) server. By building system packages, end user's can quickly create, update, and/or delete a server. These system packages build the latest source code from GitHub for the [PublicEngine](https://github.com/TheAnswer/PublicEngine) and [Core3](https://github.com/TheAnswer/Core3) projects to make a fully functioning SWGEmu server.
+This repository provides the necessary files to build unofficial RPM packages for the Star Wars Galaxies Emulator (SWGEmu) server. This uses pinned versions of git commits from GitHub for the [Core3](https://github.com/TheAnswer/Core3) and [PublicEngine](https://github.com/TheAnswer/PublicEngine) projects.
 
 
 ## RPM
 
-The RPM is only officially supported on Fedora >= 28 (64-bit).
+The RPM is only officially supported on Fedora >= 28 and RHEL/CentOS 8 (x86_64).
 
 How to create the RPM:
 ~~~
@@ -25,7 +25,7 @@ $ sudo dnf builddep ~/rpmbuild/SPECS/swgemu-server.spec
 $ rpmbuild -bb ~/rpmbuild/SPECS/swgemu-server.spec
 ~~~
 
-The compiled RPM will be available at `~/rpmbuild/RPMS/x86_64/swgemu-server-<ISO_TIMESTAMP>-<RPM_SPEC_RELEASE>.<OS_VERSION>.x86_64.rpm`.
+The compiled RPM will be available at `~/rpmbuild/RPMS/x86_64/swgemu-server*.rpm`.
 
 
 ## Server Configuration
@@ -36,11 +36,9 @@ A SWGEmu server adminsitrator's guide is provided in this repository at `rpm/SOU
 ## FAQ
 
 * There is `XYZ` problem with SWGEmu. What should I do?
-	* The source code is compiled from the latest unstable code that is in constant development. It is very likely that it is a bug in the source code. Visit [https://www.swgemu.com/bugs/](https://www.swgemu.com/bugs/) to view or open bug reports. Upstream patches can be submitted to [http://review.swgemu.com](http://review.swgemu.com). If the problem is directly related to the package creation then open an issue in this GitHub project.
-* Why is Red Hat Enterprise Linux (RHEL) not supported?
-	* SWGEmu requires at least Lua version 5.2 for support for newer functions such as `lua_version` and `luaL_traceback`. The development team recommends Lua 5.3. RHEL 7 provides Lua 5.1 by default and [will not rebase due to compatibility and stability reasons](https://bugzilla.redhat.com/show_bug.cgi?id=1437243). Replacing the system Lua library is not supported.
-* Why is Debian not supported?
-	* There is already an official SWGEmu development virtual machine called [ZonamaDev](https://github.com/Scurby/ZonamaDev). This uses Debian 9 (Stretch) as the operating system. There is also support for deploying SWGEmu to [Debian docker containers on Kubernetes](https://github.com/TheAnswer/Core3/commit/5815f8f975f899f626bf39e8283ae1040f087db7).
+	* Only open GitHub issues here if there are issues building, installing, or updating the RPM itself. Visit [https://www.swgemu.com/bugs/](https://www.swgemu.com/bugs/) to view or open bug reports. Upstream patches can be submitted to [http://review.swgemu.com](http://review.swgemu.com).
+* Why are there no Debian packages?
+	* There is already an official SWGEmu development virtual machine called [ZonamaDev](https://github.com/Scurby/ZonamaDev). This uses Debian as the operating system. There is also support for deploying SWGEmu to [Debian docker containers on Kubernetes](https://github.com/TheAnswer/Core3/commit/5815f8f975f899f626bf39e8283ae1040f087db7).
 
 
 # License

--- a/README.md
+++ b/README.md
@@ -11,14 +11,16 @@ This repository aims to provide easily available and unofficial Linux packages f
 
 ## RPM
 
-The RPM is only offiically supported on Fedora >= 25 (64-bit).
+The RPM is only officially supported on Fedora >= 28 (64-bit).
 
 How to create the RPM:
 ~~~
-$ sudo dnf install 'dnf-command(builddep)' rpm-build
+$ docker run -it --name swgemubuild fedora:28 bash
+$ sudo dnf install 'dnf-command(builddep)' git rpm-build
+$ git clone https://github.com/ekultails/swgemu-server-packages.git
 $ mkdir -p ~/rpmbuild/{BUILD,RPMS,SOURCES,SPECS,SRPMS}
-$ cp -f -r rpm/SOURCES/* ~/rpmbuild/SOURCES/
-$ cp -f -r rpm/SPECS/* ~/rpmbuild/SPECS/
+$ cp -f -r swgemu-server-packages/rpm/SOURCES/* ~/rpmbuild/SOURCES/
+$ cp -f -r swgemu-server-packages/rpm/SPECS/* ~/rpmbuild/SPECS/
 $ sudo dnf builddep ~/rpmbuild/SPECS/swgemu-server.spec
 $ rpmbuild -bb ~/rpmbuild/SPECS/swgemu-server.spec
 ~~~

--- a/rpm/SPECS/swgemu-server.spec
+++ b/rpm/SPECS/swgemu-server.spec
@@ -89,6 +89,10 @@ exit 0
 
 
 %changelog
+* Mon Jun 24 2019 Luke Short <ekultails@gmail.com 20190623-1
+- Use specific git commits for the build
+- Use the date of the latest commit for the swgemu-server version
+
 * Sat Apr 7 2018 Luke Short <ekultails@gmail.com> 5
 - Removed unnecessary dependencies
 - Rebased the Makefile's generic processor compilation patch

--- a/rpm/SPECS/swgemu-server.spec
+++ b/rpm/SPECS/swgemu-server.spec
@@ -2,8 +2,8 @@
 %define publicengine_commit 8c39f72d8cd0242d1b09a6a4d3614599e57df70a
 
 Name: swgemu-server
-Version: %{timestamp_iso}
-Release: 5%{?dist}
+Version: 20181231
+Release: 1%{?dist}
 Summary: Run a Star Wars Galaxies server with SWGEmu.
 License: GPLv3
 URL: https://github.com/ekultails/swgemu-server-packages

--- a/rpm/SPECS/swgemu-server.spec
+++ b/rpm/SPECS/swgemu-server.spec
@@ -1,4 +1,5 @@
-%define timestamp_iso %(date +"%Y%m%d")
+%define core3_commit c8872ec05aae83066271cc7226c0eb6899b936e5
+%define publicengine_commit 8c39f72d8cd0242d1b09a6a4d3614599e57df70a
 
 Name: swgemu-server
 Version: %{timestamp_iso}
@@ -6,42 +7,21 @@ Release: 5%{?dist}
 Summary: Run a Star Wars Galaxies server with SWGEmu.
 License: GPLv3
 URL: https://github.com/ekultails/swgemu-server-packages
-BuildRequires: automake findutils git gcc gcc-c++ java-1.8.0-openjdk-headless libdb-devel lua-devel make mariadb-devel pandoc
+%undefine _disable_source_fetch
+SOURCE0: https://github.com/TheAnswer/Core3/archive/%{core3_commit}.tar.gz
+SOURCE1: https://github.com/TheAnswer/PublicEngine/archive/%{publicengine_commit}.tar.gz
+BuildRequires: automake cmake findutils git gcc gcc-c++ java-1.8.0-openjdk-headless libdb-devel lua-devel make mariadb-devel pandoc
 Requires: java-1.8.0-openjdk-headless lua libdb shadow-utils
-
 
 %description
 
+%prep
+tar -x -v -f %{SOURCE0}
+tar -x -v -f %{SOURCE1}
 
 %build
-
-if [[ -d "PublicEngine" ]]; then
-	cd PublicEngine
-	git checkout master
-	# Undo any local changes.
-	git reset --hard
-	# Delete any other non-repository files such as existing compiled code.
-	git clean -f -d
-	# Pull the latest code.
-	git pull origin master
-	cd ..
-else
-	git clone --depth 1 https://github.com/TheAnswer/PublicEngine.git
-fi
-
-if [[ -d "Core3" ]]; then
-	cd Core3
-	git checkout unstable
-	git reset --hard
-	git clean -f -d
-	git pull origin unstable
-	cd ..
-else
-	git clone --depth 1 https://github.com/TheAnswer/Core3.git
-fi
-
 pushd .
-cd PublicEngine/MMOEngine
+cd PublicEngine-%{publicengine_commit}/MMOEngine
 make
 chmod 755 bin/idlc
 
@@ -61,11 +41,11 @@ popd
 
 # If the symbolic link to "MMOEngine" does not exist,
 # then create it.
-if [[ ! -h "Core3/MMOEngine" ]]; then
-	ln -s ../PublicEngine/MMOEngine Core3/MMOEngine
+if [[ ! -h "Core3-%{core3_commit}/MMOEngine" ]]; then
+	ln -s ../PublicEngine-%{publicengine_commit}/MMOEngine Core3-%{core3_commit}/MMOEngine
 fi
 
-cd Core3/MMOCoreORB
+cd Core3-%{core3_commit}/MMOCoreORB
 make config
 patch -p2 < %{_sourcedir}/Makefile_generic_x86-64_build.patch
 make rebuild
@@ -87,8 +67,8 @@ exit 0
 rm -rf %{buildroot}
 mkdir -p %{buildroot}/usr/bin/ %{buildroot}/opt/swgemu-server/doc/\
  %{buildroot}/usr/lib/systemd/system/
-cp -r Core3/MMOCoreORB %{buildroot}/opt/swgemu-server/
-cp -r PublicEngine/MMOEngine %{buildroot}/opt/swgemu-server/
+cp -r Core3-%{core3_commit}/MMOCoreORB %{buildroot}/opt/swgemu-server/
+cp -r PublicEngine-%{publicengine_commit}/MMOEngine %{buildroot}/opt/swgemu-server/
 cp %{_sourcedir}/swgemu-server.service %{buildroot}/usr/lib/systemd/system/
 find %{buildroot} -name ".git*" -delete
 pandoc %{_sourcedir}/readme.md > %{buildroot}/opt/swgemu-server/doc/readme.html

--- a/rpm/SPECS/swgemu-server.spec
+++ b/rpm/SPECS/swgemu-server.spec
@@ -1,8 +1,8 @@
-%define core3_commit c8872ec05aae83066271cc7226c0eb6899b936e5
-%define publicengine_commit 8c39f72d8cd0242d1b09a6a4d3614599e57df70a
+%define core3_commit b8776cc66a0c96e6d78494e3f389e48f43d469d1
+%define publicengine_commit 1bbdb8a182a9e44bc23f1d972ac254c1ca98db03
 
 Name: swgemu-server
-Version: 20181231
+Version: 20190623
 Release: 1%{?dist}
 Summary: Run a Star Wars Galaxies server with SWGEmu.
 License: GPLv3
@@ -10,7 +10,7 @@ URL: https://github.com/ekultails/swgemu-server-packages
 %undefine _disable_source_fetch
 SOURCE0: https://github.com/TheAnswer/Core3/archive/%{core3_commit}.tar.gz
 SOURCE1: https://github.com/TheAnswer/PublicEngine/archive/%{publicengine_commit}.tar.gz
-BuildRequires: automake cmake findutils git gcc gcc-c++ java-1.8.0-openjdk-headless libdb-devel lua-devel make mariadb-devel pandoc
+BuildRequires: automake cmake findutils git gcc gcc-c++ java-1.8.0-openjdk-headless libatomic libdb-devel lua-devel make mariadb-devel pandoc
 Requires: java-1.8.0-openjdk-headless lua libdb shadow-utils
 
 %description
@@ -48,7 +48,9 @@ fi
 cd Core3-%{core3_commit}/MMOCoreORB
 make config
 patch -p2 < %{_sourcedir}/Makefile_generic_x86-64_build.patch
-make rebuild
+make config
+make cleanidl
+make -j 4 build-cmake
 cd %{_builddir}
 # This will find and delete all files that contain a word and
 # end with the file extension ".cpp" or ".h" (source files).

